[WTF] Extract and use forEachSetBit in BitVector and BitSet

Constellation opened the pull request:
#### b54f40faba449f186c14912cebc01a6c02d4710f
<pre>
[WTF] Extract and use forEachSetBit in BitVector and BitSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=280149">https://bugs.webkit.org/show_bug.cgi?id=280149</a>
<a href="https://rdar.apple.com/136445959">rdar://136445959</a>

Reviewed by Yijia Huang.

This patch extracts the implementation of forEachSetBit from BitSet and
share it with BitVector.

* Source/WTF/wtf/BitSet.h:
(WTF::WordType&gt;::forEachSetBit const):
* Source/WTF/wtf/BitVector.h:
(WTF::BitVector::forEachSetBit const):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::forEachSetBit):

Canonical link: <a href="https://commits.webkit.org/284109@main">https://commits.webkit.org/284109@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fbeb34d1605e667920c25998c34fa1085922b50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68326 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20985 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72392 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19471 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19287 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54558 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12964 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43638 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59010 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35022 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40306 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16430 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17828 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61444 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62263 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16779 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74085 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67574 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12297 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16042 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62011 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59089 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62031 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9951 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3569 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89353 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10420 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43519 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15803 "Found 9 new JSC stress test failures: wasm.yaml/wasm/function-tests/many-arguments-to-function.js.default-wasm, wasm.yaml/wasm/fuzz/memory.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-collect-continuously, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-eager-jettison, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit, wasm.yaml/wasm/stress/tail-call-simple.js.wasm-slow-memory, wasm.yaml/wasm/stress/tail-call.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44593 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45787 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44335 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WTF/wtf/BitSet.h:
```diff
@@ -397,11 +397,11 @@ inline constexpr bool BitSet<bitSetSize, WordType>::subsumes(const BitSet& other
-template<typename Func>
-ALWAYS_INLINE constexpr void BitSet<bitSetSize, WordType>::forEachSetBit(const Func& func) const
+template<typename WordType, typename Func>
+ALWAYS_INLINE constexpr void forEachSetBit(std::span<const WordType> bits, const Func& func)
```
These should probably be moved to StdLibExtras.h

I think these could also be:
```
template<std::integral WordType>
constexpr void forEachSetBit(std::span<const WordType> bits, const Invocable<IterationStatus(size_t)> auto& func)
...


// Convenience wrapper for when you want to iterate the whole set.
template<std::integral WordType>
constexpr void forEachSetBit(std::span<const WordType> bits, const Invocable<void(size_t)> auto& func)
{
    forEachBitSet(bits, [&] (size_t i) {
        func(i);
        return IterationStatus::Continue;
    }
}
```
which would make the API a bit clearer. It might also be possible to implement the 0 start in terms of the `startIndex` version below? It seems like the compiler should optimize away the startIndex if it's a constant 0?

Constellation replied:
Changed.
But for now, I don't want to change `startIndex` version separate as this patch's intent is not accidentally regressing performance.

Constellation replied:
Ah, unfortunately, rather using Invocable here causes interface non-cleaner because compiler failed to identify which function it would like to use, and emitting ambiguous error. And it requires us to write all type signature explicitly in the caller side, and we would like to avoid this.

So, here, I'll just move this to StdLibExtras.h

kmiller68 commented on Source/WTF/wtf/BitSet.h:
```diff
@@ -397,11 +397,11 @@ inline constexpr bool BitSet<bitSetSize, WordType>::subsumes(const BitSet& other
+ALWAYS_INLINE constexpr void forEachSetBit(std::span<const WordType> bits, const Func& func)
 {
-    for (size_t i = 0; i < words; ++i) {
+    constexpr size_t wordSize = sizeof(WordType) * 8;
```
Nit: `sizeof(WordType) * CHAR_BIT`

Constellation replied:
Changed.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[WTF] Extract and use forEachSetBit in BitVector and BitSet
https://bugs.webkit.org/show_bug.cgi?id=280149
rdar://136445959

Reviewed by Yijia Huang.

This patch extracts the implementation of forEachSetBit from BitSet and
share it with BitVector.

* Source/WTF/wtf/BitSet.h:
(WTF::WordType>::forEachSetBit const):
* Source/WTF/wtf/BitVector.h:
(WTF::BitVector::forEachSetBit const):
* Source/WTF/wtf/StdLibExtras.h:
(WTF::forEachSetBit):

Canonical link: https://commits.webkit.org/284109@main
